### ISSUE TITLE
Infinite loop detected ... and caused by Style/RedundantBegin

### ISSUE THREAD

**Issue by tsrivishnu** [rubocop/rubocop#10116]
I have the following code

```ruby
# app/controllers/application_controller.rb
  helper_method def user_job_roles_characteristics
    @ _user_job_roles_characteristics ||=
      begin
        if !context_user.guest?
          ::Characteristics::ListService.call(service_context)
            .output
            .characteristics
        else
          []
        end
      end
  end
```

When I run rubocop with `-A` option, it fails with `Infinite loop detected ...` error.

--------

## Expected behavior

The offences to be autocorrected.

## Actual behavior

An error was raised:
```bash
$ bundle exec rubocop --debug -A app/controllers/application_controller.rb
For /app: configuration from /app/.rubocop.yml
configuration from /app/.bundle/ruby/2.7.0/gems/rubocop-gitlab-security-0.1.1/config/default.yml
configuration from /app/.bundle/ruby/2.7.0/gems/rubocop-gitlab-security-0.1.1/config/default.yml
Default configuration from /app/.bundle/ruby/2.7.0/gems/rubocop-1.14.0/config/default.yml
configuration from /app/.bundle/ruby/2.7.0/gems/rubocop-rails-2.11.3/config/default.yml
configuration from /app/.bundle/ruby/2.7.0/gems/rubocop-rails-2.11.3/config/default.yml
configuration from /app/.bundle/ruby/2.7.0/gems/rubocop-rspec-2.2.0/config/default.yml
configuration from /app/.bundle/ruby/2.7.0/gems/rubocop-rspec-2.2.0/config/default.yml
Inheriting configuration from /app/.rubocop_todo.yml
.rubocop.yml: GitlabSecurity/PublicSend:Exclude overrides the same parameter in .rubocop_todo.yml
Inspecting 1 file
Scanning /app/app/controllers/application_controller.rb
W

Offenses:

app/controllers/application_controller.rb:91:5: C: [Corrected] Style/MultilineMemoization: Wrap multiline memoization blocks in begin and end.
    @_user_job_roles_characteristics ||= ...
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
app/controllers/application_controller.rb:92:7: C: [Corrected] Style/RedundantBegin: Redundant begin block detected.
      begin
      ^^^^^
app/controllers/application_controller.rb:93:8: C: [Corrected] Layout/IndentationWidth: Use 2 (not 3) spaces for indentation.
          ::Characteristics::ListService.call(service_context)
       ^^^
app/controllers/application_controller.rb:94:9: C: [Corrected] Layout/IndentationWidth: Use 2 (not 1) spaces for indentation.
         ::Characteristics::ListService.call(service_context)
        ^
app/controllers/application_controller.rb:96:9: C: [Corrected] Layout/ElseAlignment: Align else with if.
        else
        ^^^^
app/controllers/application_controller.rb:97:8: C: [Corrected] Layout/ElseAlignment: Align else with if.
       else
       ^^^^
app/controllers/application_controller.rb:98:8: C: [Corrected] Layout/IndentationWidth: Use 2 (not 3) spaces for indentation.
          []
       ^^^
app/controllers/application_controller.rb:98:9: W: [Corrected] Layout/EndAlignment: end at 98, 8 is not aligned with if at 92, 7.
        end)
        ^^^
app/controllers/application_controller.rb:98:9: C: [Corrected] Layout/IndentationWidth: Use 2 (not 1) spaces for indentation.
         []
        ^
app/controllers/application_controller.rb:99:1: C: [Corrected] Layout/TrailingWhitespace: Trailing whitespace detected.
app/controllers/application_controller.rb:99:8: W: [Corrected] Layout/EndAlignment: end at 99, 7 is not aligned with if at 93, 8.
       end
       ^^^
app/controllers/application_controller.rb:101:1: C: [Corrected] Layout/EmptyLinesAroundMethodBody: Extra empty line detected at method body end.

0 files inspected, 12 offenses detected, 12 offenses corrected
Infinite loop detected in /app/app/controllers/application_controller.rb and caused by Style/RedundantBegin -> Layout/ElseAlignment, Layout/EndAlignment, Layout/IndentationWidth, Layout/TrailingWhitespace, Style/MultilineMemoization -> Layout/ElseAlignment, Layout/EmptyLinesAroundMethodBody, Layout/EndAlignment, Layout/IndentationWidth, Style/RedundantBegin -> Layout/IndentationWidth, Style/RedundantBegin
/app/.bundle/ruby/2.7.0/gems/rubocop-1.14.0/lib/rubocop/runner.rb:299:in `check_for_infinite_loop'
/app/.bundle/ruby/2.7.0/gems/rubocop-1.14.0/lib/rubocop/runner.rb:282:in `block in iterate_until_no_changes'
/app/.bundle/ruby/2.7.0/gems/rubocop-1.14.0/lib/rubocop/runner.rb:281:in `loop'
/app/.bundle/ruby/2.7.0/gems/rubocop-1.14.0/lib/rubocop/runner.rb:281:in `iterate_until_no_changes'
/app/.bundle/ruby/2.7.0/gems/rubocop-1.14.0/lib/rubocop/runner.rb:250:in `do_inspection_loop'
/app/.bundle/ruby/2.7.0/gems/rubocop-1.14.0/lib/rubocop/runner.rb:130:in `block in file_offenses'
/app/.bundle/ruby/2.7.0/gems/rubocop-1.14.0/lib/rubocop/runner.rb:155:in `file_offense_cache'
/app/.bundle/ruby/2.7.0/gems/rubocop-1.14.0/lib/rubocop/runner.rb:129:in `file_offenses'
/app/.bundle/ruby/2.7.0/gems/rubocop-1.14.0/lib/rubocop/runner.rb:120:in `process_file'
/app/.bundle/ruby/2.7.0/gems/rubocop-1.14.0/lib/rubocop/runner.rb:101:in `block in each_inspected_file'
/app/.bundle/ruby/2.7.0/gems/rubocop-1.14.0/lib/rubocop/runner.rb:100:in `each'
/app/.bundle/ruby/2.7.0/gems/rubocop-1.14.0/lib/rubocop/runner.rb:100:in `reduce'
/app/.bundle/ruby/2.7.0/gems/rubocop-1.14.0/lib/rubocop/runner.rb:100:in `each_inspected_file'
/app/.bundle/ruby/2.7.0/gems/rubocop-1.14.0/lib/rubocop/runner.rb:86:in `inspect_files'
/app/.bundle/ruby/2.7.0/gems/rubocop-1.14.0/lib/rubocop/runner.rb:47:in `run'
/app/.bundle/ruby/2.7.0/gems/rubocop-1.14.0/lib/rubocop/cli/command/execute_runner.rb:26:in `block in execute_runner'
/app/.bundle/ruby/2.7.0/gems/rubocop-1.14.0/lib/rubocop/cli/command/execute_runner.rb:52:in `with_redirect'
/app/.bundle/ruby/2.7.0/gems/rubocop-1.14.0/lib/rubocop/cli/command/execute_runner.rb:25:in `execute_runner'
/app/.bundle/ruby/2.7.0/gems/rubocop-1.14.0/lib/rubocop/cli/command/execute_runner.rb:17:in `run'
/app/.bundle/ruby/2.7.0/gems/rubocop-1.14.0/lib/rubocop/cli/command.rb:11:in `run'
/app/.bundle/ruby/2.7.0/gems/rubocop-1.14.0/lib/rubocop/cli/environment.rb:18:in `run'
/app/.bundle/ruby/2.7.0/gems/rubocop-1.14.0/lib/rubocop/cli.rb:65:in `run_command'
/app/.bundle/ruby/2.7.0/gems/rubocop-1.14.0/lib/rubocop/cli.rb:72:in `execute_runners'
/app/.bundle/ruby/2.7.0/gems/rubocop-1.14.0/lib/rubocop/cli.rb:41:in `run'
/app/.bundle/ruby/2.7.0/gems/rubocop-1.14.0/exe/rubocop:12:in `block in <top (required)>'
/usr/local/lib/ruby/2.7.0/benchmark.rb:308:in `realtime'
/app/.bundle/ruby/2.7.0/gems/rubocop-1.14.0/exe/rubocop:12:in `<top (required)>'
/app/.bundle/ruby/2.7.0/bin/rubocop:23:in `load'
/app/.bundle/ruby/2.7.0/bin/rubocop:23:in `<top (required)>'
/usr/local/lib/ruby/2.7.0/bundler/cli/exec.rb:63:in `load'
/usr/local/lib/ruby/2.7.0/bundler/cli/exec.rb:63:in `kernel_load'
/usr/local/lib/ruby/2.7.0/bundler/cli/exec.rb:28:in `run'
/usr/local/lib/ruby/2.7.0/bundler/cli.rb:476:in `exec'
/usr/local/lib/ruby/2.7.0/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
/usr/local/lib/ruby/2.7.0/bundler/vendor/thor/lib/thor/invocation.rb:127:in `invoke_command'
/usr/local/lib/ruby/2.7.0/bundler/vendor/thor/lib/thor.rb:399:in `dispatch'
/usr/local/lib/ruby/2.7.0/bundler/cli.rb:30:in `dispatch'
/usr/local/lib/ruby/2.7.0/bundler/vendor/thor/lib/thor/base.rb:476:in `start'
/usr/local/lib/ruby/2.7.0/bundler/cli.rb:24:in `start'
/usr/local/lib/ruby/gems/2.7.0/gems/bundler-2.1.4/libexec/bundle:46:in `block in <top (required)>'
/usr/local/lib/ruby/2.7.0/bundler/friendly_errors.rb:123:in `with_friendly_errors'
/usr/local/lib/ruby/gems/2.7.0/gems/bundler-2.1.4/libexec/bundle:34:in `<top (required)>'
/app/bin/bundle:104:in `load'
/app/bin/bundle:104:in `<main>'
Finished in 2.008201500007999 seconds
ERROR: 2
```

## Steps to reproduce the problem

Use the following rubocop config and the above mentioned code
```yaml
inherit_from: .rubocop_todo.yml

require:
  - rubocop-gitlab-security
  - rubocop-rails
  - rubocop-rspec
  - rubocop-i18n

AllCops:
  Exclude:
    - 'db/schema.rb'
    - 'node_modules/**/*'
  NewCops: enable # Enable new cops when we upgrade by default

Style/GlobalVars:
  AllowedVariables:
    - $kafka

Layout/MultilineMethodCallIndentation:
  EnforcedStyle: indented
Style/AndOr:
  EnforcedStyle: conditionals
Style/NumericPredicate:
  EnforcedStyle: comparison

Style/HashEachMethods:
  Enabled: true
Style/HashTransformKeys:
  Enabled: true
Style/HashTransformValues:
  Enabled: true
Style/NegatedIf:
  Enabled: false
Style/StringLiterals:
  EnforcedStyle: double_quotes


GitlabSecurity/JsonSerialization:
  Enabled: false
GitlabSecurity/PublicSend:
  Exclude:
    - "spec/**/*_spec.rb"
Rails/CreateTableWithTimestamps:
  Enabled: false
Rails/Delegate:
  Enabled: false # It is sometimes nice to not do +delegate+ to favour readability
Rails/ShortI18n:
  Enabled: true
  EnforcedStyle: aggressive
Rails/I18nLocaleAssignment:
  Enabled: true
  Include:
    - config/**/*.rb
    - app/**/*.rb
    - spec/**/*.rb
    - misc/**/*.rb
    - test/**/*.rb
    - lib/**/*.rb

Rails/NotNullColumn:
  # in case of an +add_reference+, this doesn't make sense.
  # We can let an exception happen.
  # Setting a default value doens't make sense in this case
  Enabled: false
Rails/UnknownEnv:
  Environments:
    - production
    - staging
    - development
    - test
Rails/WhereNot:
  Enabled: false

RSpec/MessageSpies:
  # This is so that we enforce Four-Phase Test https://thoughtbot.com/blog/four-phase-test
  # pattern. This isn't set on stone. We can revisit this setting anytime with
  # strong arguments, for or against.
  EnforcedStyle: have_received
RSpec/DescribedClass:
  Enabled: false

Lint/AmbiguousBlockAssociation:
  Exclude:
    - "spec/**/*_spec.rb"

# See: https://github.com/puppetlabs/rubocop-i18n#usage
I18n/RailsI18n:
  Enabled: true
I18n/GetText:
  Enabled: false
I18n/RailsI18n/DecorateString:
  Enabled: false
```

## RuboCop version

```
$ bundle exec rubocop -V
1.14.0 (using Parser 3.0.2.0, rubocop-ast 1.11.0, running on ruby 2.7.1 x86_64-linux)
  - rubocop-rails 2.11.3
  - rubocop-rspec 2.2.0
```

**Comment by dvandersluis** [comment]
I do not get an infinite loop with your code and configuration. You're using an older version of rubocop however so it's could have been fixed already. Can you try updating to the latest version?
